### ISSUE TITLE
SDPA Memory Efficient Attention is producing NaNs for certain mask patterns.

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#109517]
### Summary
For certain mask patterns the memory_efficeint_attention kernel which is called via SDPA is producing NaNs. 

### Reproducing Script 
I was able to validate this on an a100 on nightly

``` Python
import torch


from torch.backends.cuda import sdp_kernel

def from_file():
    load_dct = torch.load('/home/drisspg/meta/scripts/sdp_tests/w.pt')
    q = load_dct['query'].contiguous()
    k = load_dct['key'].contiguous()
    v = load_dct['value'].contiguous()
    mask = load_dct['attn_mask'].contiguous()
    print(f"Max Mask Value: {mask.max()}")
    print(f"Min Mask Value: {mask.min()}")
    return q,k,v,mask

def from_sizes():
    seq_len_size = 441
    shape = (32, 12, seq_len_size, 64)
    mask_shape = (32, 12, seq_len_size, seq_len_size)
    q = torch.randn(shape, device='cuda')
    k = torch.randn(shape, device='cuda')
    v = torch.randn(shape, device='cuda')
    mask = torch.zeros(mask_shape, device='cuda')
    mask[0:,0:,0:,0:128] = torch.finfo(torch.float32).min
    return q,k,v,mask


def main():
    q,k,v,mask = from_file()
    # q,k,v,mask = from_sizes()

    q_mem,k_mem,v_mem = q.detach().clone(),k.detach().clone(),v.detach().clone()
    mask_mem = mask.detach().clone()

    q.requires_grad_(True)
    k.requires_grad_(True)
    v.requires_grad_(True)
    mask.requires_grad_(True)

    q_mem.requires_grad_(True)
    k_mem.requires_grad_(True)
    v_mem.requires_grad_(True)
    mask_mem.requires_grad_(True)

    out_grad = torch.rand_like(q_mem)
    with sdp_kernel(enable_mem_efficient=False):
        out_math = torch.nn.functional.scaled_dot_product_attention(q,k,v,attn_mask=mask)
        out_math.backward(out_grad)
    with sdp_kernel(enable_mem_efficient=True):
        out_mem = torch.nn.functional.scaled_dot_product_attention(q_mem,k_mem,v_mem,attn_mask=mask_mem)
        out_mem.backward(out_grad)

    assert not torch.any(torch.isnan(out_math)), "Math output contains NaN"
    assert not torch.any(torch.isnan(out_mem)), "MemEff output contains NaN"

    torch.testing.assert_close(out_math, out_mem)
    torch.testing.assert_close(q.grad, q_mem.grad)
    torch.testing.assert_close(k.grad, k_mem.grad)
    torch.testing.assert_close(v.grad, v_mem.grad)
    torch.testing.assert_close(mask.grad, mask_mem.grad)

    print(f"Max Diff out = {torch.max(torch.abs(out_math - out_mem))}")
    print(f"Max Diff q = {torch.max(torch.abs(q.grad - q_mem.grad))}")
    print(f"Max Diff k = {torch.max(torch.abs(k.grad - k_mem.grad))}")
    print(f"Max Diff v = {torch.max(torch.abs(v.grad - v_mem.grad))}")
    print(f"Max Diff mask = {torch.max(torch.abs(mask.grad - mask_mem.grad))}")
if __name__ == "__main__":
    main()
```

### Problem
The attn_mask being used for this input is a certain pattern of 0 and "-inf".  My suspicion is that a block is being completely masked and this is causing Nans. I have observed this behavior when working with the triton implementation of FlashAttn.  

I have verified this not the same issue as having a completely masked out row.  As well contiguous on the inputs does not have any effect. 
This is hitting the cause where we pad and slice, since the seq_len is not a multiple of 16: https://github.com/pytorch/pytorch/blob/5d5990fc49260cfd717df629017720109c89f1d4/aten/src/ATen/native/transformers/attention.cpp#L487-L504

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by drisspg** [comment]
The input tensors can be found here: https://filetransfer.io/data-package/dBondS0V#link

**Comment by drisspg** [comment]
@danthe3rd I am not sure if you have encountered issues like this before

**Comment by malfet** [comment]
Adding triage review for silient correctness issue

**Comment by drisspg** [comment]
@danthe3rd 
https://github.com/pytorch/pytorch/pull/109620

I updated to the above PR here above and still seeing the error

**Comment by drisspg** [comment]
Also, it appears we already do a -inf check for `mi` here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/cuda/mem_eff_attention/kernel_forward.h#L1195-L1201

**Comment by drisspg** [comment]
Okay I was suspecting that this was not the same issue as: https://github.com/pytorch/pytorch/issues/103749

But after much `printf` in the kernel I do indeed suspect this is the same issue of having a row completely masked out.

I printed the attn_mask[0,0,392]:
```Shell
(Pdb) p mask_mem[0,0,392]
tensor([-3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38, -3.4028e+38,
        -3.4028e+38]
   ```
   And it is entirely masked. So with that in mind I am going to close this issue.
